src/Lykke.Job.CandlesHistoryWriter.Services/Candles/CandlesCacheInitalizationService.cs

```diff
@@ -28,27 +30,48 @@ public class CandlesCacheInitalizationService : ICandlesCacheInitalizationServic
             ICandlesHistoryRepository candlesHistoryRepository,
             int amountOfCandlesToStore)
         {
-            _log = log;
-            _assetPairsManager = assetPairsManager;
-            _clock = clock;
-            _candlesCacheService = candlesCacheService;
-            _candlesHistoryRepository = candlesHistoryRepository;
+            _log = log ?? throw new ArgumentNullException(nameof(log));
+            _assetPairsManager = assetPairsManager ?? throw new ArgumentNullException(nameof(assetPairsManager));
+            _clock = clock ?? throw new ArgumentNullException(nameof(clock));
+            _candlesCacheService = candlesCacheService ?? throw new ArgumentNullException(nameof(candlesCacheService));
+            _candlesHistoryRepository = candlesHistoryRepository ?? throw new ArgumentNullException(nameof(candlesHistoryRepository));
             _amountOfCandlesToStore = amountOfCandlesToStore;
+
+            InitializationState = CacheInitializationState.Idle;
         }
 
         public async Task InitializeCacheAsync()
         {
-            await _log.WriteInfoAsync(nameof(CandlesCacheInitalizationService), nameof(InitializeCacheAsync), null, "Caching candles history...");
+            // Depending on cache invalidation period and on asset pairs amount, there may be a case when
+            // the invalidation timer fires before the cache loading has stopped. This will be a signal 
+            // to skip timer-based invalidation.
+            if (InitializationState == CacheInitializationState.InProgress)
```

Comment:
This is not thread safe check

Reply:
Not safe in general. But here we have static field of a singleton service. It's something like a semaphore, but is simplier.
Could you explane with more details in private messages?

Reply:
It's not a semaphore at all. Checking and updating of the InitializationState - is not atomic.
For instance if timer in the RedisCacheCaretaker.cs will be triggered right after the start, and cache initialization will be started by the startup manager, then we will be faced with race condition. Both threads could see that InitializationState is CacheInitializationState.Idle then both will update InitializationState to the CacheInitializationState.InProgress and run cache initialization in parallel.